cache.py

```diff
@@ -0,0 +1,19 @@
+from weakref import WeakValueDictionary
+
+from entities import *
+from utils import *
+
+
+class Cache:
+    def __init__(self):
+        self._cache = WeakValueDictionary()
+
+    def add_or_update(self, transaction):
+        self._cache[transaction.header_id] = transaction
+
+    def retrieve(self, header_id):
+        if header_id in self._cache:
```

Comment:
I think it would be considered more pythonic, and easier to read imo, if this was defensive, like

```
if header_id not in self._cache:
    return None
    
if self._cache[header_id].deleted:
    # maybe remove from cache?
    return None

return self._cache[header_id]
```